webapp/packages/plugin-data-viewer/src/ResultSetDataSource.ts

```diff
@@ -5,19 +5,50 @@
  * Licensed under the Apache License, Version 2.0.
  * you may not use this file except in compliance with the License.
  */
+import { makeObservable, observable } from 'mobx';
+
 import type { IServiceInjector } from '@cloudbeaver/core-di';
+import type { ITask } from '@cloudbeaver/core-executor';
 import type { AsyncTaskInfoService, GraphQLService } from '@cloudbeaver/core-sdk';
 
 import { DatabaseDataSource } from './DatabaseDataModel/DatabaseDataSource';
 import type { IDatabaseResultSet } from './DatabaseDataModel/IDatabaseResultSet';
 
+interface ResultSetDataSourceState {
+  cancelLoadTotalCountTasks: Map<string, ITask<number>>;
+}
+
 export abstract class ResultSetDataSource<TOptions> extends DatabaseDataSource<TOptions, IDatabaseResultSet> {
+  readonly cancelLoadTotalCountTasks: Map<string, ITask<number>>;
+
   constructor(
     readonly serviceInjector: IServiceInjector,
     protected graphQLService: GraphQLService,
     protected asyncTaskInfoService: AsyncTaskInfoService,
   ) {
     super(serviceInjector);
+
+    this.cancelLoadTotalCountTasks = new Map();
+
+    makeObservable<ResultSetDataSourceState>(this, {
+      cancelLoadTotalCountTasks: observable.shallow,
+    });
+  }
+
+  async cancelLoadTotalCount(resultIndex: number) {
+    const result = this.getResult(resultIndex);
+
+    if (!result?.id) {
+      throw new Error('Result id must be provided');
+    }
+
+    const task = this.cancelLoadTotalCountTasks.get(result.id);
+
+    if (!task?.cancelled) {
+      task?.cancel();
+    }
+
+    this.cancelLoadTotalCountTasks.delete(result.id);
```

Comment:
But it will be deleted when the task finishes (you don't need to duplicate it here and spread add/delete logic between different scopes)